deltacat/compute/stats/basic.py

```diff
@@ -0,0 +1,33 @@
+from typing import Dict, Set, Tuple
+
+from deltacat.compute.stats.models.stats_result import StatsResult
+from deltacat.compute.stats.types import StatsType
+from deltacat.compute.stats.utils.io import discover_deltas_stats
+
+from deltacat.storage import PartitionLocator
+from deltacat.storage import interface as unimplemented_deltacat_storage
+
+
+def collect(
+        source_partition_locator: PartitionLocator,
+        delta_stream_position_range_set: Set[Tuple[int, int]],  # TODO: R&D on RangeSet impl
+        stat_types_to_collect: Set[StatsType],
+        stat_results_s3_bucket: str = None,  # TODO: Decouple DeltaCAT from S3-based paths
+        deltacat_storage=unimplemented_deltacat_storage
+) -> Dict[int, StatsResult]:
+    delta_stream_range_stats: Dict[int, StatsResult] = {}
+    for range_pair in delta_stream_position_range_set:
+        begin_pos, end_pos = range_pair
+        delta_stats: Dict[int, StatsResult] = discover_deltas_stats(
+            source_partition_locator,
+            begin_pos,
+            end_pos,
+            stat_types_to_collect,
+            stat_results_s3_bucket,
+            deltacat_storage
+        )
+        delta_stream_range_stats.update(delta_stats)
```

Comment:
The current implementation will synchronously collect stats for each interval while only parallelizing stats collection across all deltas in that interval, but we will also want to parallelize collection of all non-overlapping intervals. 

To do this, we may also need to apply a workaround from https://github.com/ray-project/ray/issues/3644 to ensure that we don't spin up too many workers when running nested tasks. One way to do this is to roll out the existing task into multiple sequential tasks to remove the need to make nested @ray.remote calls, which may look something like this:
```python
    discover_deltas_pending = []
    for range_pair in delta_stream_position_range_set:
        begin_pos, end_pos = range_pair
        promise: Dict[int, StatsResult] = read_cached_delta_stats.remote(
            source_partition_locator,
            begin_pos,
            end_pos,
            stat_types_to_collect,
            stat_results_s3_bucket,
            deltacat_storage,
        )
        discover_deltas_pending.append(promise)
     delta_stats_pending = []
     while len(discover_deltas_pending):
        ready, discover_deltas_pending = ray.wait(discover_deltas_pending)
        delta_stats, cache_miss_delta_locators = ray.get(ready[0])
        # assume delta_stats will be None/empty if not cached
        # assume cache_miss_delta_locators contains the delta locator for every cache miss
        if cache_miss_delta_locators:
            promise = _get_delta_stats.remote(
                source_partition_locator,
                start_pos,
                end_pos,
                deltacat_storage,
            )
            delta_stats_pending.append(promise)
        else:
          delta_stream_range_stats.update(delta_stats)
      cache_stats_pending = []
      while len(delta_stats_pending):
        ready, delta_stats_pending = ray.wait(delta_stats_pending)
        promise = cache_pending_stats.remote(ready)
        cache_stats_pending.append(promise)
      # synchronously wait for all stats to be cached
      delta_stats_list = ray.get(cache_stats_pending)
      for delta_stats in delta_stats_list:
        delta_stream_range_stats.update(delta_stats)
```

Reply:
Updated the workflow by rolling up remote Ray calls and avoiding nested remote calls